src/main/java/frc/robot/subsystems/DriveSubsystem.java

```diff
@@ -0,0 +1,124 @@
+package frc.robot.subsystems;
+
+
+import com.ctre.phoenix.motorcontrol.can.WPI_TalonSRX;
+import com.ctre.phoenix.motorcontrol.can.WPI_VictorSPX;
+import com.ctre.phoenix6.configs.Pigeon2Configuration;
+import com.ctre.phoenix6.hardware.Pigeon2;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
+import edu.wpi.first.wpilibj.smartdashboard.Field2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.RobotMap;
+
+public class DriveSubsystem extends SubsystemBase {
+    private final WPI_TalonSRX leftFrontMotor;
+    private final WPI_VictorSPX rightFrontMotor;
+    private final WPI_VictorSPX leftBackMotor;
+    private final WPI_TalonSRX rightBackMotor;
+    private final Field2d field2d;
+    private final DifferentialDriveOdometry differentialDriveOdometry;
+    private Pigeon2 pigeon2;
+
+    public DriveSubsystem() {
+        leftBackMotor = new WPI_VictorSPX(RobotMap.DRIVE_LEFT_BACK_MOTOR_ID);
+        rightBackMotor = new WPI_TalonSRX(RobotMap.DRIVE_RIGHT_BACK_MOTOR_ID);
+        leftFrontMotor = new WPI_TalonSRX(RobotMap.DRIVE_LEFT_FRONT_MOTOR_ID);
+        rightFrontMotor = new WPI_VictorSPX(RobotMap.DRIVE_RIGHT_FRONT_MOTOR_ID);
+        pigeon2 = new Pigeon2(RobotMap.PIGEON_ID);
+        leftBackMotor.configFactoryDefault();
+        rightBackMotor.configFactoryDefault();
+        leftFrontMotor.configFactoryDefault();
+        rightFrontMotor.configFactoryDefault();
+        pigeon2.getConfigurator().apply(new Pigeon2Configuration());
+
+        leftFrontMotor.setInverted(true);
+        leftBackMotor.setInverted(true);
+        leftFrontMotor.setSensorPhase(true);
+
+        this.field2d = new Field2d();
+
+        differentialDriveOdometry = new DifferentialDriveOdometry(
+                new Rotation2d(getAngleDegrees()),
+                getLeftDistancePassedMeters(),
+                getRightDistancePassedMeters()
+        );
+        field2d.setRobotPose(differentialDriveOdometry.getPoseMeters().getX(), differentialDriveOdometry.getPoseMeters().getY(), differentialDriveOdometry.getPoseMeters().getRotation());
+
+        initialize();
+
+
+    }
+
+    public Field2d getField2d() {
+        return field2d;
+    }
+
+    public double getLeftDistancePassedMeters() {
+        return leftFrontMotor.getSelectedSensorPosition() / RobotMap.TALON_ENCODER_PPR * RobotMap.DRIVE_WHEEL_RADIUS * 2 * Math.PI;
+    }
+
+    public double getRightDistancePassedMeters() {
+        return rightBackMotor.getSelectedSensorPosition() / RobotMap.TALON_ENCODER_PPR * RobotMap.DRIVE_WHEEL_RADIUS * 2 * Math.PI;
+    }
+
+    public void initialize() {
```

Comment:
this is only used inside this class. Make `private`

Reply:
private

Reply:
Why do you call it there though? This methods resets the encoders and pigeon. As such, calling it in the command will reset the encoders and pigeon. 

This has a problematic affect of having the encoders and pigeon back at 0, losing all the information they accumulated. We have the odometery which use this information. Having the sensor lose their values will break the `odometery.update`, as it calculates distance travel relative to the last encoder values.

Reply:
ohhhhhh, yes yes youre right, ill change it to private